src/gias3/learning/PCA.py

```diff
@@ -65,54 +66,76 @@ def __init__(self, mean=None, weights=None, modes=None, SD=None, projectedWeight
+        def _load_pickle(file):
+            with open(file, 'rb') as f:
+                try:
+                    return pickle.load(f)
+                except UnicodeDecodeError:
+                    return pickle.load(f, encoding='latin1')
 
-    def load(self, filename):
-        # return self.loadz(filename)
-        try:
-            return self.loadz(filename)
-        except:
+        def _load_npz(file):
+            try:
+                return numpy.load(file, allow_pickle=True)
+            except OSError:
+                return numpy.load(file, encoding='bytes', allow_pickle=True)
+
+        def _load_shelve(file):
+            return shelve.open(file, 'r')
+
+        for method in [_load_pickle, _load_npz, _load_shelve]:
             try:
-                return self.loadPickle(filename)
-            except EOFError:
-                return self.loadShelve(filename)
-
-    def saveShelve(self, filename):
-        s = shelve.open(filename + '.pc')
-        s['mean'] = self.mean
-        s['weights'] = self.weights
-        s['modes'] = self.modes
-        s['SD'] = self.SD
-        s['sizes'] = self.sizes
-        s['projectedWeights'] = self.projectedWeights
-        s.close()
-        return filename + '.pc'
-
-    def savePickle(self, filename):
-        s = {'mean': self.mean, 'weights': self.weights, 'modes': self.modes, 'SD': self.SD, 'sizes': self.sizes, 'projectedWeights': self.projectedWeights}
-        with open(filename + '.pc', 'w') as f:
-            pickle.dump(s, f)
-
-        return filename + '.pc'
-
-    def savez(self, filename):
-        if os.path.splitext(filename)[1].lower() != '.pc':
-            filename += '.pc'
-
-        numpy.savez(
-            filename,
-            mean=self.mean,
-            weights=self.weights,
-            modes=self.modes,
-            SD=self.SD,
-            sizes=self.sizes,
-            projectedWeights=self.projectedWeights
-        )
+                data = method(file_path)
+
+                # Assign variables from data.
+                byte_keys = b'mean' in data
+                self.mean = data[b'mean'] if byte_keys else data['mean']
+                self.weights = data[b'weights'] if byte_keys else data['weights']
+                self.modes = data[b'modes'] if byte_keys else data['modes']
+                self.SD = data[b'SD'] if byte_keys else data['SD']
+                if self.SD is not None:
+                    if len(self.SD.shape) != 0:
+                        self.sdNorm = True
+                try:
+                    self.projectedWeights = data[b'projectedWeights'] if byte_keys else data['projectedWeights']
+                except KeyError:
+                    self.projectedWeights = None
+                try:
+                    self.sizes = data[b'sizes'] if byte_keys else data['sizes']
+                except KeyError:
+                    self.sizes = None
+
+                # Load successful.
+                return
```

Comment:
Ii would wrap all this in a method `_load` and then do my catch all exception block around just the `data = method(file_path)` line.

Reply:
@hsorby, I've just made some changes. Not sure if it's quite what you wanted. If I put the `try` block inside the `_load` method then that would need to return or throw something to indicate if that method in the loop was successful. Let me know what you think.